state/watcher.go

```diff
@@ -496,6 +496,33 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
 	return newRelationLifeSuspendedWatcher(backend, members, filter, nil)
 }
 
+// WatchModelMachinesLifeAndAgentStartTimes watches the machines in the model
+// for changes to the Life or AgentStartTime fields.
+func (st *State) WatchModelMachinesLifeAndAgentStartTimes() StringsWatcher {
+	docStates := make(map[string]string)
+	return newCollectionWatcher(st, colWCfg{
```

Comment:
This is clever but it feels really dirty! The filter function is normally used to determine whether this is an id we're interested in, not to query the db and compare the current state to the old one. 

I can see one filter function that does query the db, `makeControllerIdFilter`, but it's not fetching the specific record that's being notified, it's getting the current list of controllers so it can look at status doc ids. That does locking around the state because the filter function can be called from different goroutines so we were seeing a race.

I feel like this would be better written as a watcher with a loop over a WatchCollectionWithFilter - but man, looking at the implementation of lifecycleWatcher I can see why you wouldn't want to. So maybe just change the structure so it locks around updates to docStates? Or write it in the same style as `makeControllerIdFilter`?

Reply:
IIRC another example of reading the doc in the filter callback is the machine address hash changed watcher.

After our quick chat today, I will change the code to use a similar batching scheme as the lifecycle watcher which should protect us from hammering the DB.

Reply:
See extended commit message in https://github.com/juju/juju/pull/10902/commits/6da33b0aae89a0f4bd562e63cf722a3211be246e for details on how I modified the previous watcher implementation.